pkg-manager/resolve-dependencies/src/getCatalogSnapshots.ts

```diff
@@ -0,0 +1,66 @@
+import { type CatalogSnapshots } from '@pnpm/lockfile-types'
+import { type ResolvedDirectDependency } from './resolveDependencyTree'
+import { type Catalogs } from '@pnpm/catalogs.types'
+
+export function getCatalogSnapshots (
+  resolvedDirectDeps: readonly ResolvedDirectDependency[],
+  catalogsConfig: Catalogs = {},
+  prevSnapshots: CatalogSnapshots = {}
+): CatalogSnapshots {
+  const nextSnapshots: CatalogSnapshots = {}
+
+  // Note that "snapshotUpdates" only contains the catalog lookups relevant to a
+  // filtered install. This is by design.
+  //
+  // To avoid wiping away previous catalog snapshots that weren't involved in a
+  // filtered install, we'll need to merge these updates in with the
+  // "prevSnapshots" from the existing wanted lockfile.
+  const snapshotUpdates = getCatalogSnapshotsForResolvedDeps(resolvedDirectDeps)
```

Comment:
Resolution will never happen on a subset of projects. I have changed this recently. So this could be simplified, no?

Reply:
Pushed up commit https://github.com/pnpm/pnpm/pull/8221/commits/31416680dabc4de4342fda27fd47bba8e7b5320d, which contains this simplification. Thank you for catching. 🥳

I also changed one of the tests to ensure stale catalog snapshots **do not** persist in the `pnpm-lock.yaml` file since we can now reliably tell if any catalog entry is still used.